### ISSUE TITLE
 Shadowed memory variable in function body by named returns statement

### ISSUE THREAD

**Issue by mojtaba-eshghie** [ethereum/solidity#13435]
## Description
In solidity, it is possible to specify the name of the returned variable from function as well as its value type. The presense of this feature cancels need to both declaring the variable in function body as well as having a `return` statement. However, if the developer mistakenly declares the variable in the function body again, the newly declared variable will not shadow the previously declared variable in function declaration statement. This leads to the function returning the default initial value for named return variable instead of the intended value that is computed in function body since it pertains to another variable only with the same name. The current solidity compiler (0.8.15) does not raise an error, and only shows a warning: `Warning: This declaration shadows an existing declaration.` for the decleration of the variable in function body. This is not correct since the declared variable in function declaration statement will shadow the variable in function body not the other way around.

## Environment

- Compiler version: 0.8.15
- Target EVM version (as per compiler settings): default for Remix
- Framework/IDE (e.g. Truffle or Remix): Remix
- EVM execution environment / backend / blockchain client: Remix vm
- Operating system: - 

## Steps to Reproduce

In the getMax function of the following code the initial value of max (in the named return statement in function declaration) will shadow the value that is assigned later in function body.

```
// SPDX-License-Identifier: MIT
pragma solidity >=0.6.0 <=0.9.0;

library ArrayUtils {
    function getMax(uint[] calldata a) public pure returns (uint max) {
        uint max = 0;
        uint max_index = 0;
        for (uint i = 0; i < a.length; i++) {
            if (a[i] >= max) {
                max_index = i;
                max = a[i];
            }
        }
    }

    function range(uint l) internal pure returns (uint[] memory r) {
        r = new uint[](l);      
        for (uint i = 0; i < l; i++) {
            r[i] = i;
        }
    }
}


contract ArraysUser {
    using ArrayUtils for *;

    function getMax(uint l) public pure returns (uint max) {
        max = ArrayUtils.range(l).getMax();
    }
}
```

**Comment by Amxx** [comment]
> This is not correct since the declared variable in function declaration statement will shadow the variable in function body not the other way around.

When you perform an operation in the function body, such a `max = a[i];`, then `max` is a reference to the `uint` declared in the function body, and not to the variable declared in return.

The fact that max in the function body is what you operate on make the max in the return inaccessible. That is why we say the max in the function body shadows the max in the return. AFAIK,here  "shadows" can means "to hide". 

Therefore I think the warning is correct.

**Comment by Marenz** [comment]
I just tested this with a minimal test case and couldn't confirm your problem:

```
  Contract:
    contract C {
    function foo() public returns (int a) {
    int a;

    a = 11;
    }
    }

  Expected result:
  // foo() -> 11

  Obtained result:
  // foo() -> 0
```

From your description you we should see a return value of `11` now if the assignment were to refer to the named return variable. As it does return `0`, it means it referenced the local variable as expected...

Closing for now. Feel free to add more comments/reopen if you feel this is wrong.
